### PR TITLE
Added Passthrough option for exposing upstream UI while ensuring multi-tenancy isolation.

### PR DESCRIPTION
**Pull request by bwplotka**
Also: Added example, and moved Frederic's example to kube-rbac-proxy dir.

This option is extremely useful for multi-tenancy without requiring Grafana, allowing to use Thanos, Prometheus UIs etc. Might be unsafe, if we miss certain APIs that's why passthrough is for non API endpoints and disabled by default.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: injectproxy/rules_test.go [hunk 1]
```diff
@@ -658,7 +658,7 @@ func TestRules(t *testing.T) {
 		t.Run(fmt.Sprintf("%s=%s", proxyLabel, tc.labelv), func(t *testing.T) {
 			m := newMockUpstream(tc.upstream)
 			defer m.Close()
-			r := NewRoutes(m.url, proxyLabel)
+			r := NewRoutes(m.url, proxyLabel,false)
```
Comment (s-urbaniak) [inline injectproxy/rules_test.go:661]: as per CI failure:
```
-			r := NewRoutes(m.url, proxyLabel,false)
+			r := NewRoutes(m.url, proxyLabel, false)
```

**Comment by s-urbaniak** [comment]
I think this is a nice (although admittedly dangerous) feature! The only thing I am wondering though if this logic cannot be handled by nginx (or another proxy) natively, where `/api` requests would be dispatched to prom-label-proxy, and all other requests directly to the upstream target, be it Prometheus or Thanos, but having it in Go code is definitely more readable/maintainable.

I am curious about thoughts from others, cc @paulfantom @simonpasquier @lilic @brancz

**Comment by bwplotka** [comment]
Yes, I thought about it, however, this would make it quite hard to use with nginx/caddy/envoy. Reasons:

* You are cannot isolate upstream from ingress, as some nginx has to access both prom-label-proxy and upstream for certain requests
* It's very special knowledge of what paths should be enforced, what is not. All is manually specified and very fragile. It would awesome to have only one well-maintained place like this, not many.

File: examples/caddy-port-based/Caddyfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+reverse_proxy 127.0.0.1:8080
+
+}
+
+:8082 {
+
+rewrite * ?{query}&job=prometheus
+reverse_proxy 127.0.0.1:8080
+
+}
```
Comment (kakkoyun) [inline examples/caddy-port-based/Caddyfile:13]:
```suggestion
}

```

**Comment by bwplotka** [comment]
Fixed (:

**Comment by s-urbaniak** [comment]
From my side this LGTM :+1: As this contribution has bigger ramifications though I'd love to have another set of eyes from the other maintainers.

**Comment by brancz** [comment]
I don't think this is safe to do, the targets page and configuration page may contain somewhat sensitive data.

**Comment by bwplotka** [comment]
With other PR, those will return "not implemented" return code so that's ok
- until we implement tenancy for those.

Kind Regards,
Bartek Płotka (@bwplotka)

**Comment by bwplotka** [comment]
Oh actually I read just now `the targets page and configuration page may contain somewhat sensitive data`. This makes sense.

I believe you are right. Full pass through is too dangerous. I would propose something else: path allowlist for passthrough with safe default of UI resources only. WDYT? 🤗 
If that makes sense, will update this PR.

Przemek, can you provide more info what exactly paths it would be necessary (and safe) to pass through in your case?

**Comment by kprzemek** [comment]
@bwplotka I will try to explain it simply ;-)

I am doing a large PoC for one of the banks in my country.
In fact, they only need pasthrough for one purpose - setting the label on the namespace (isolation to a specific tenant in the query). Access to the Targets and Configuration page is absolutely not necessary.

An example from my Caddy (prom-label-proxy: 10.11.20.22:29090):
11.10.20.22:9092 {
    rewrite *? {query} & namespace = project1
    reverse_proxy 10.11.20.22:29090
}
11.10.20.22:9093 {
    rewrite *? {query} & namespace = project2
    reverse_proxy 10.11.20.22:29090
}

    Thanks to this, they can do DataSource in Grafana, for example: http://10.11.20.22:9092 - project1-DS or http://10.11.20.22:9093 - project2-DS. Isolation at the proxy level will ensure that in Grafana, a user with a specific DataSource [IP: Port for example: http://10.11.20.22:9092] will not be able to see someone else's (other namespace) data.

Everything else can go like through a revese proxy:
11/10/20.22:9091 {
    reverse_proxy 10/11/20.22:9090
}

In this case, it will be the DataSource address in Grafana for Administrators (for example: http://10.11.20.22:9091 - admin-DS) who have unlimited access to all namespaces / tenents.

Prom-label-proxy seems the most appropriate. By the way, it is connected to the Prometheus community. Everything under one roof ;-)

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,8 @@
+## Unreleased
+
+* [FEATURE] Add -non-api-path-passthrough flag (disabled by default), which allows
+ exposing all non API resources from upstream without enforcing (e.g Prometheus UI).
```
Comment (brancz) [inline CHANGELOG.md:4]: Let's add a note how this is safe in regards to for example the targets and configuration pages.

**Review by brancz — COMMENTED** [review]
Is the non-react UI entirely removed in Prometheus? Otherwise this will potentially leak information that was available through the legacy UI.

**Comment by bwplotka** [comment]
> Is the non-react UI entirely removed in Prometheus?

It's not. But we could block access to that (or allow users to do so).

Still if we allowlist pass through for UI resources (JS, CSS) only (le'ts say) even if there is button for Target, Config etc - proxy will not route to that.

**Comment by brancz** [comment]
Could we at least put the same not-implemented thing in there for the old UI as we have here for non explicit API endpoints, and add documentation that if people use this flag Prometheus has to be >= version x.y.z? Otherwise I feel that we will be creating a bunch of vulnerable Prometheus instances.

**Comment by bwplotka** [comment]
Sure, I will propose something safe.

**Comment by kprzemek** [comment]
So ok. 
When will the passthrough be in the main branch? ;-)
For our PoC, I need it in the official prom-label-proxy branch.

**Comment by bwplotka** [comment]
Will try today / this week

**Comment by bwplotka** [comment]
Sorry for delay, I addressed all comments.

cc @brancz @s-urbaniak @kprzemek PTAL 🤗

**Comment by brancz** [comment]
lgtm

**Review by kakkoyun — APPROVED** [review]

**Comment by kprzemek** [comment]
@bwplotka thank you for this commit.
I try tu use it, and I have some problem. Guys please help me. Maybe it's my browser's problem?
I try to open 3 paths: -unsafe-passthrough-paths="/graph,/stores,/new"
but when I try to open it in browser I have only one URL OK - /stores.
Sorry for the ignorance, but I'm stuck .
TGraph
TNewGraph
TStore

**Comment by bwplotka** [comment]
It works, however you need to find out list of paths that you need to show. Check F12 console and network tab and see which ones are throwing 404 and pass them in this option if it's safe to do so from isolation point of view 🤗

**Comment by kprzemek** [comment]
Works.
Thank you. It was /api/v1... I add them to -unsafe-passthrough-paths and voilà ;-)

**Pull request merged**